File: 101/exercises/easy6/cuteangles.rb

MINUTES_PER_DEGREE = 60
SECONDS_PER_MINUTE = 60

def dms(float)
  float_within_360 = float.abs - ((float.abs / 360).to_i * 360)

  degrees, remainder = float_within_360.divmod 1
  minutes, remainder = (remainder * MINUTES_PER_DEGREE).divmod 1
  seconds = remainder * SECONDS_PER_MINUTE

  sign = float.negative? ? '-' : ''
  format(%(#{sign}%d°%02d'%02d"), degrees, minutes, seconds)
end

## Tests

p dms(30) == %(30°00'00")
p dms(76.73) == %(76°43'48")
p dms(254.6) == %(254°35'59")
p dms(93.034773) == %(93°02'05")
p dms(0) == %(0°00'00")
p dms(360) == %(360°00'00") || dms(360) == %(0°00'00")
p dms(440) == %(80°00'00")
p dms(-76.73) == %(-76°43'48")

File: 210/small-problems/easy4/cute-angles.js
function dms(angle) {
  var MINUTES_PER_DEGREE = 60
  var SECONDS_PER_MINUTE = 60

  // make sure angle is in (0..360)
  angle = Math.abs(angle) - Math.floor(angle / 360) * 360;

  var degrees   = Math.floor(angle);
  var remainder = MINUTES_PER_DEGREE * (angle - degrees);
  var minutes   = Math.floor(remainder);
  var seconds   = Math.floor(SECONDS_PER_MINUTE * (remainder - minutes));

  return String(degrees) + '˚' + format(minutes) + '\'' + format(seconds) + '"';
}

function format(number) {
  var str = String(number);

  if (str.length == 1) {
    str = '0' + str;
  }
  
  return str;
}

console.log(dms(30));           // 30°00'00"
console.log(dms(76.73));        // 76°43'48"
console.log(dms(254.6));        // 254°35'59"
console.log(dms(93.034773));    // 93°02'05"
console.log(dms(0));            // 0°00'00"
console.log(dms(360));          // 360°00'00" or 0°00'00"
